### PR TITLE
improve tester implementation

### PR DESCRIPTION
**Pull request by frairon**
Reimplementation of the tester to  fix all issues (see issue #145)

### CODE REVIEW

File: examples/4-tests/example_test.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+	// consume a message
+	gkt.ConsumeString("input", "key", "some message")
+
+	// ensure the message was received
+	ensure.DeepEqual(t, receivedMessage, "some message")
+}
+
+// Scenario (2)
+//
```
Comment (juandes) [inline examples/4-tests/example_test.go]: Silly comment. You have this unused comment line :)

File: tester/messagetracker.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
+package tester
+
+// MessageTracker tracks message offsets for each topic for convenient
+// 'expect message x to be in topic y' in unit tests
+type MessageTracker struct {
```
Comment (db7) [inline tester/messagetracker.go]: I would suggest bind the MessageTracker to a topic. That will simplify the interface.

File: tester/messagetracker.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	return &MessageTracker{
+		t:                t,
+		nextMessageIndex: make(map[string]int),
+		tester:           tester,
+	}
+}
+
+// NextMessage returns the next message from the topic since the last time this
+// function was called (or MoveToEnd)
+func (mt *MessageTracker) NextMessage(topic string) (string, interface{}, bool) {
```
Comment (db7) [inline tester/messagetracker.go]: suggestion: use `Next()` instead

File: tester/messagetracker.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+	decoded, err := mt.tester.codecForTopic(topic).Decode(msgRaw)
+	if err != nil {
+		mt.t.Fatalf("Error decoding message: %v", err)
+	}
+	return key, decoded, true
+}
+
+// NextMessageRaw returns the next message in passed topic
+func (mt *MessageTracker) NextMessageRaw(topic string) (string, []byte, bool) {
```
Comment (db7) [inline tester/messagetracker.go]: suggestion: use `NextRaw()` instead

File: tester/messagetracker.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+		return "", nil, false
+	}
+	msg := q.message(mt.nextMessageIndex[topic])
+
+	mt.nextMessageIndex[topic]++
+	return msg.key, msg.value, true
+}
+
+// ExpectEmpty ensures the topic does not contain more messages
+func (mt *MessageTracker) ExpectEmpty(topic string) {
```
Comment (db7) [inline tester/messagetracker.go]: This is not necessary since once can call `Next()` until it returns false

File: tester/messagetracker.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+func (mt *MessageTracker) MoveAllToEnd() {
+	for topic := range mt.tester.topicQueues {
+		mt.nextMessageIndex[topic] = int(mt.tester.queueForTopic(topic).hwm)
+	}
+}
+
+// ExpectEmit ensures a message exists in passed topic and key. The message may be
+// inspected/unmarshalled by a passed expecter function.
+// DEPRECATED: This function is only to get some compatibility and should be removed in future
+func (mt *MessageTracker) ExpectEmit(topic string, key string, expecter func(value []byte)) {
```
Comment (db7) [inline tester/messagetracker.go]: I'd remove this method.

File: tester/tester.go [hunk 7]
```diff
@@ -20,35 +23,84 @@ type Codec interface {
...
+	q, exists := km.topicQueues[topic]
+	if !exists {
+		panic(fmt.Errorf("No queue for topic %s", topic))
+	}
+	return q
+}
+
+// CreateMessageTracker creates a message tracker that starts tracking
+// the messages from the end of the current queues
+func (km *Tester) NewMessageTrackerFromEnd() *MessageTracker {
```
Comment (db7) [inline tester/tester.go]: suggestion: `NewMessageTracker(topic string)`

File: tester/messagetracker.go (hunk 5 above)
Comment (frairon) [reply tester/messagetracker.go]: actually I'd leave the message, as it does more than just checking we're at the end. Also having a shortcut for the alternative loop that we would have to replae in all projects would be helpful.
but I'd rename it to `ExpectAtEnd` to have it consistend with `MoveToEnd`.
What do you think?

Comment (db7) [reply tester/messagetracker.go]: What about replacing `ExpectAtEnd()` with something like `mt.Offset() == mt.HWM()` ?

and  `MoveToEnd()` replacing with `mt.Seek(mt.HWM())`?

Comment (frairon) [reply tester/messagetracker.go]: agreed

**Pull request merged**